### ISSUE TITLE
Incorrect type for the default argument of `Mapping.pop()` when used in logical expression

### ISSUE THREAD

**Issue by vberlier** [microsoft/pyright#4062]
In the following example, Pyright v1.1.275 resolves the type of the default argument as `Literal[True]` instead of `None`. This only happens when the call to `pop()` is part of a logical `and/or` expression.

```py
from typing import Mapping

a: Mapping[str, int] = {}
b = True and a.pop("thing", None)
# error: Argument of type "None" cannot be assigned to parameter "__default" of type "int | _T@pop" in function "pop"
#     Type "None" cannot be assigned to type "int | Literal[True]"
#       Type "None" cannot be assigned to type "int"
#       Type "None" cannot be assigned to type "Literal[True]" (reportGeneralTypeIssues)
```

**Comment by erictraut** [comment]
This is due to [bidirectional type inference](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#bidirectional-type-inference-expected-types), which is applied to compound expressions joined by `and` and `or` operators. The assumption is that for such expressions, the left-hand type and right-hand type are going to be the same. This is a good assumption in most code, and it helps produce the correct type inference for the RHS in most cases.

This is admittedly a heuristic, and it can lead to false positives in some rare cases like the one you've highlighted above. I think we've made the right tradeoff here. Fixing the false positive in your code would introduce false positives in much more common code patterns.

I'll note that mypy uses the same heuristic here and produces the same error as pyright.

I can offer a couple of simple workarounds:

```python
# Produce a bool value for both the LHS and RHS
b = True and a.pop("thing", None) is not None
```

```python
# Use a temporary variable
c = a.pop("thing", None)
b = True and c
```

**Comment by vberlier** [comment]
I see, thanks for the explanation.
